### PR TITLE
ENH: Upgrade ITK from C++14 to C++17

### PR DESCRIPTION
**Pull request by N-Dekker**
Follow-up to pull request https://github.com/InsightSoftwareConsortium/ITK/pull/2563 commit daec0fdd703b1bca16aeb07bc4e0a01101da950a "ENH: Upgrade ITK from C++11 to C++14"

Suggested by Dženan Zukić (@dzenanz) 😃 at https://github.com/InsightSoftwareConsortium/ITK/pull/3933#discussion_r1128770848 Still to be discussed.

More information about C++17:
- https://en.cppreference.com/w/cpp/17
- https://en.wikipedia.org/wiki/C%2B%2B17 
- https://isocpp.org/files/papers/p0636r0.html
- https://www.cppstories.com/2017/11/cpp17summary/
- etc...!

Discussion "Shall we start using C++17 at the master branch?" opened at https://discourse.itk.org/t/shall-we-start-using-c-17-at-the-master-branch/5828

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -33,9 +33,9 @@ endforeach()
 include(${CMAKE_CURRENT_SOURCE_DIR}/CMake/ITKInitializeCXXStandard.cmake)
 include(${CMAKE_CURRENT_SOURCE_DIR}/CMake/ITKInitializeBuildType.cmake)
 
-# ==== Define language standard configurations requiring at least c++14 standard
-if(CMAKE_CXX_STANDARD EQUAL "98" OR CMAKE_CXX_STANDARD LESS "14")
-   message(FATAL_ERROR "C++98 to C++11 are no longer supported in ITK version 5.3 and greater.")
+# ==== Define language standard configurations requiring at least c++17 standard
+if(CMAKE_CXX_STANDARD EQUAL "98" OR CMAKE_CXX_STANDARD LESS "17")
+   message(FATAL_ERROR "C++98 to C++14 are no longer supported in ITK version 5.3 and greater.")
```
Comment (dzenanz) [inline CMakeLists.txt:38]: ITK version 5.3 -> ITK version 5.4

Comment (N-Dekker) [reply CMakeLists.txt]: @dzenanz Thanks, addressed with the most recent [force-push](https://github.com/InsightSoftwareConsortium/ITK/compare/099c723c5d94bf12f2a188ad185042e3dbcc8067..0ec405cafff5ab1bfed0cb5c494755c151a3ca81) 😃

**Comment by N-Dekker** [comment]
Still need to figure out how to upgrade the Python wrapping from C++14 to C++17! Any clue? Looking at [Darwin-Build8553-PR3969-CXX_STANDARD-from-14-to-17-Python](https://open.cdash.org/viewBuildError.php?buildid=8558018):

> cd /.../s-build/Wrapping/Modules/ITKCommon && env SDKROOT=/Applications/Xcode_13.2.1.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk MACOSX_DEPLOYMENT_TARGET=11.7 /.../s-build/Wrapping/Generators/CastXML/castxml/bin/castxml -o /.../s-build/Wrapping/castxml_inputs/ITKCommonBase.xml --castxml-gccxml --castxml-start _wrapping_ --castxml-cc-gnu "(" /Applications/Xcode_13.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -Wall -Wcast-align -Wdisabled-optimization -Wextra -Wformat=2 -Winvalid-pch -Wno-format-nonliteral -Wpointer-arith -Wshadow -Wunused -Wwrite-strings -funit-at-a-time -Wno-strict-overflow -Wno-deprecated -Wno-invalid-offsetof -Wno-undefined-var-template -Woverloaded-virtual -std=c++14 ")" -w -c @/.../s-build/Wrapping/castxml_inputs/.castxml.inc /.../s-build/Wrapping/castxml_inputs/ITKCommonBase.cxx


It still says `-std=c++14`. That should be _-std=c++17_, right?

Similarly, [Linux-Build8480-PR3969-CXX_STANDARD-from-14-to-17-Python](https://open.cdash.org/viewBuildError.php?buildid=8557868) says:

> cd /.../s-build/Wrapping/Modules/ITKCommon && /.../s-build/Wrapping/Generators/CastXML/castxml/bin/castxml -o /.../s-build/Wrapping/castxml_inputs/ITKCommonBase.xml --castxml-gccxml --castxml-start _wrapping_ --castxml-cc-gnu "(" /usr/bin/c++ -Wall -Wcast-align -Wdisabled-optimization -Wextra -Wformat=2 -Winvalid-pch -Wno-format-nonliteral -Wpointer-arith -Wshadow -Wunused -Wwrite-strings -funit-at-a-time -Wno-strict-overflow -Wno-deprecated -Wno-invalid-offsetof -Woverloaded-virtual -Wstrict-null-sentinel -msse2 -msse2 -std=c++14 ")" -w -c @/.../s-build/Wrapping/castxml_inputs/.castxml.inc /.../s-build/Wrapping/castxml_inputs/ITKCommonBase.cxx

Also still says `-std=c++14`! What to do?

**Comment by issakomi** [comment]
> Also still says -std=c++14!

In _Wrapping/macro_files/itk_auto_load_submodules.cmake_, [here](https://github.com/InsightSoftwareConsortium/ITK/blob/bdc523419f3fe180addc122136a6e6912209b47c/Wrapping/macro_files/itk_auto_load_submodules.cmake#L25)
`        set(_castxml_cc_flags "${_castxml_cc_flags} ${CMAKE_CXX14_EXTENSION_COMPILE_OPTION}")`

But I'm just guessing, not sure.

**Comment by N-Dekker** [comment]
> In _Wrapping/macro_files/itk_auto_load_submodules.cmake_, [here](https://github.com/InsightSoftwareConsortium/ITK/blob/bdc523419f3fe180addc122136a6e6912209b47c/Wrapping/macro_files/itk_auto_load_submodules.cmake#L25)

Thank you @issakomi ! That's very helpful  😃 ! Just [force-pushed](https://github.com/InsightSoftwareConsortium/ITK/compare/0ec405cafff5ab1bfed0cb5c494755c151a3ca81..561d522b8f7cf4de12a67a58a5a0b33a53da75bc)!

**Comment by dzenanz** [comment]
Maybe rebase on new master now that silence PR was merged?

File: Documentation/SupportedCompilers.md [hunk 2]
```diff
@@ -1,21 +1,21 @@
-ITK requires a compiler with C++14 support.
+ITK requires a compiler with C++17 support.
 
 # Visual Studio
 * VS2015 and earlier: **NOT supported**
-* MSVC toolset v141 (first shipped with VS2017): supported
+* MSVC toolset v141 (first shipped with VS2017): supported from VS2017 version 15.7
 * MSVC toolset v142 (first shipped with VS2019): supported
```
Comment (dzenanz) [inline Documentation/SupportedCompilers.md:6]: Should we also explicitly list VS2022?

File: Modules/Core/Common/include/itkMacro.h [hunk 3]
```diff
@@ -151,19 +151,15 @@ namespace itk
 #endif
 
 /*
- * ITK only supports MSVC++ 14.0 and greater
- * MSVC++ 14.0 _MSC_VER == 1900 (Visual Studio 2015 version 14.0)
- * MSVC++ 14.1 _MSC_VER == 1910 (Visual Studio 2017 version 15.0)
- * MSVC++ 14.11 _MSC_VER == 1911 (Visual Studio 2017 version 15.3)
- * MSVC++ 14.12 _MSC_VER == 1912 (Visual Studio 2017 version 15.5)
+ * ITK only supports MSVC++ 14.14 and greater
  * MSVC++ 14.13 _MSC_VER == 1913 (Visual Studio 2017 version 15.6)
```
Comment (dzenanz) [inline Modules/Core/Common/include/itkMacro.h:155]: Should this line be removed too?

**Review by dzenanz — COMMENTED** [review]
Good progress!

File: Modules/Core/Common/include/itkMacro.h (hunk 3 above)
Comment (N-Dekker) [reply Modules/Core/Common/include/itkMacro.h]: Thanks again, addressed with the latest [force-push](https://github.com/InsightSoftwareConsortium/ITK/compare/57fd767dc14f8b1d1d6d9ecf44cf48454707e649..de7124eacb3466359c1ba529f2e83b4d183ba74f)

**Comment by dzenanz** [comment]
There seems to be consensus support in the [forum discussion](https://discourse.itk.org/t/shall-we-start-using-c-17-at-the-master-branch/5828). Please review here too.

**Comment by bradking** [comment]
Do: check

**Comment by bradking** [comment]
It's more fallout from the [GitHub host key change](https://github.blog/2023-03-23-we-updated-our-rsa-ssh-host-key/).  It should be fixed now.

**Review by thewtex — APPROVED** [review]

**Comment by seanm** [comment]
Hmm, unfortunate that such a big change was made with such a short discussion period. I just discovered this last night.

Requiring Xcode 10 consequently drops support for macOS 10.10, 10.11, and 10.12 because Xcode 10 can only run on macOS 10.13.

Dropping 10.10 support is not so bad because 10.11 supports all the same Macs as 10.10.  I'll retire the Rogue16 nightly cdash build consequently.

Dropping 10.11 support though means dropping support for various Mac models because not all Macs that can run 10.11 can run 10.12, as 10.12 dropped support for various models.  The newest Xcode that can be run on 10.11 is Xcode 8.2.1, which is what Rogue7 runs.

Dropping 10.12 support is not so bad because 10.13 supports all the same Macs as 10.11.

I still have customers with older Macs (that are thus stuck on 10.11), and so it's been nice to build and run nightly ITK tests with my corresponding bots.  I suppose I don't necessarily need to *build* on these old OSes though: hopefully an ITK built on a newer macOS/Xcode can still result in an executable that can still *run* on these older OSes.  I imagine that will be less about C++ itself but more about what macOS 10.11's libc++ supports or doesn't.  I'll update my 10.13 bot (Rogue22) to reduce the `CMAKE_OSX_DEPLOYMENT_TARGET` to `10.11` and let's see what that gives...

**Pull request merged**